hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+      } else if (instr->IsUncondBranchImm()) {
+        imm >>= vixl::kInstructionSizeLog2;
+        if (!is_int26(imm) || far.count(instr)) {
+          dest_block.setFrontier(dest);
+          dest_count--;
+
+          vixl::MacroAssembler a { dest_block };
+          a.SetScratchRegisters(vixl::NoReg, vixl::NoReg);
+          auto const tmp = rVixlScratch0;
+          a.Mov(tmp, instr->ImmPCOffsetTarget());
+          a.Br(tmp);
+          a.SetScratchRegisters(rVixlScratch0, rVixlScratch1);
+
+          dest_count += (dest_block.frontier() - dest) / kInstructionSize;
+          is_relative = false;
+        }
+      }
+
+      // Update offset it was converted from relative to absolute
+      if (is_relative) {
+        instr2->SetImmPCOffsetTarget(instr->ImmPCOffsetTarget());
+      } else {
+        rewrites.insert(instr);
+      }
+    }
+
+    if (instr->IsCondBranchImm() ||
+        instr->IsUncondBranchImm()) {
+      jmp_dest = target;
+    }
+    return true;
+  } else {
+    return false;
+  }
+}
+
+/*
+ * Returns true if the instruction(s) contain an immediates
+ *   MOV/MOVK
+ *   LDR (literal) - Not supported as it's highly unlikely to have a
+ *                   19 bit address
+ */
+size_t relocateImmediateHelper(RelocationInfo& rel,
+                               CodeBlock& dest_block, TCA src, TCA dest,
+                               size_t& src_count, size_t& dest_count,
+                               InstrSet &rewrites) {
+  Instruction* instr = Instruction::Cast(src);
+  Instruction* instr2 = Instruction::Cast(dest);
+
+  if (instr->IsMovz()) {
+    const auto rd = instr->Rd();
+    uint64_t target = instr->ImmMoveWide() << (16 * instr->ShiftMoveWide());
+    auto next = instr->NextInstruction();
+    while (next->IsMovk()) {
+      if (next->Rd() == rd) {
+        target |= next->ImmMoveWide() << (16 * next->ShiftMoveWide());
+      }
+      next = next->NextInstruction();
+    }
+
+    TCA adjusted = rel.adjustedAddressAfter((TCA)target);
+    if (!adjusted) { adjusted = (TCA)target; }
+    int imm = Instruction::Cast(adjusted) - instr2;
+    bool is_absolute = true;
+
+    /*
+     * If the next instruction after a sequence of MOV/MOVK uses the
+     * target, see if the entire sequence can be converted to a
+     * single PC relative instruction. Supported transformations include:
+     *   MOV/MOVK, LDR to LDR literal - Not yet implemented
+     *   MOV/MOVK, BR<L> to B<L>
+     *   MOV/MOVK absolute to ADR
+     */
+    if (next->Mask(LoadStoreUnsignedOffsetMask) == LDR_x_unsigned
+        && next->Rt() == rd) {
+      // Only transform if the MOV/MOVK sequence def'd a vixl scratch,
+      // otherwise we run the risk of not def'ing a live register.
+      auto const dst = vixl::Register(rd, 64);
+      if (dst.Is(rVixlScratch0) || dst.Is(rVixlScratch1)) {
+        if (is_int19(imm)) {
+          // TODO(CDE): Support MOV/MOVK, LDR to LDR literal.
+          not_implemented();
```

Comment:
The optimal optimizations we can definitely skip; but this pass should be correct, even on workloads not exercised by our unit tests.